### PR TITLE
[SPARK-24402] [SQL] Optimize `In` expression when only one element in the collection or collection is empty

### PR DESCRIPTION
**Pull request by dbtsai**
## What changes were proposed in this pull request?

Two new rules in the logical plan optimizers are added.

1. When there is only one element in the **`Collection`**, the
physical plan will be optimized to **`EqualTo`**, so predicate
pushdown can be used.

```scala
    profileDF.filter( $"profileID".isInCollection(Set(6))).explain(true)
    """
      |== Physical Plan ==
      |*(1) Project [profileID#0]
      |+- *(1) Filter (isnotnull(profileID#0) && (profileID#0 = 6))
      |   +- *(1) FileScan parquet [profileID#0] Batched: true, Format: Parquet,
      |     PartitionFilters: [],
      |     PushedFilters: [IsNotNull(profileID), EqualTo(profileID,6)],
      |     ReadSchema: struct<profileID:int>
    """.stripMargin
```

2. When the **`Collection`** is empty, and the input is nullable, the
logical plan will be simplified to

```scala
    profileDF.filter( $"profileID".isInCollection(Set())).explain(true)
    """
      |== Optimized Logical Plan ==
      |Filter if (isnull(profileID#0)) null else false
      |+- Relation[profileID#0] parquet
    """.stripMargin
```

TODO:

1. For multiple conditions with numbers less than certain thresholds,
we should still allow predicate pushdown.
2. Optimize the **`In`** using **`tableswitch`** or **`lookupswitch`**
when the numbers of the categories are low, and they are **`Int`**,
**`Long`**.
3. The default immutable hash trees set is slow for query, and we
should do benchmark for using different set implementation for faster
query.
4. **`filter(if (condition) null else false)`** can be optimized to false.

## How was this patch tested?

Couple new tests are added.

### CODE REVIEW

**Comment by dbtsai** [comment]
+@cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: This can be moved inside `case expr @ In(v, list) if expr.inSetConvertible`.

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: This has a bug when the Literal is a struct. See the test failure: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91218/testReport/org.apache.spark.sql/SQLQueryTestSuite/sql/

Comment (dbtsai) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: As you suggested, I'll move into `case expr @ In(v, list) if expr.inSetConvertible`. 

Yes, I saw the test failure. The same code was passing the build in my another PR. https://github.com/apache/spark/pull/21416/commits/1332406d7f4ca7a9a4a85338f758430ecc334ff8  I will debug it tomorrow.

Thanks,

**Comment by dbtsai** [comment]
@maropu I didn't do a full performance benchmark. I believe the performance gain can be from predicate pushdown when only one element in the set. This can be a lot.

I forgot which one, I remember in impala or drill,  they allow multiple predicates to be pushed down, and I believe this can be a win in many cases.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: When `list.length == 1`, we don't need to create `ExpressionSet`.

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: This is too minor, I'd like to keep the current code and not break the code flow.

**Comment by cloud-fan** [comment]
LGTM, pending jenkins

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: nit: size => length
because we use `length` in the previous `if`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: nit: In line 235 the comment 
```// newList.length == list.length```
can be updated as 
```// newList.length == list.length && newList.length > 1```

**Review by gengliangwang — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Sounds good.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: This will fail, since the schema mismatches when the data type is struct. The test cases were added a few days ago. https://github.com/apache/spark/pull/21425

**Comment by gatorsmile** [comment]
@HyukjinKwon The code has a bug. Please read the discussion before you trigger the retest.

**Comment by HyukjinKwon** [comment]
@gatorsmile, can I just see if it the build passes against the latest build?

**Comment by gatorsmile** [comment]
@HyukjinKwon All the involved reviewers will get a ping. This is annoying to see many pings within one hour, right? My suggestion is to read the comments before triggering the test

**Comment by HyukjinKwon** [comment]
This should get updated, right? I should give a ping here anyway. I wonder why triggering retesting so matters for some PRs.

Probably, you are mostly talking about "ok to test" that I left for the old PRs which gave pings for guys. In this case, I think the root causes is that we started to block the Jenkins tests for the reason I don't know. This already gives a lot of pings though so far. If the author is willing to update its PR, then it's a blocker again to ask Jenkins build.

**Comment by HyukjinKwon** [comment]
Also, see the stale PRs and see what we have delayed to take a look. Please consider partly this is what we should have checked and reviewed earlier, and/or partly authors haven't updated their PRs. Both cases need pings, right? This is what we should do but we couldn't. Mostly these were inactive for more than a month.

**Comment by gatorsmile** [comment]
@HyukjinKwon Thank you for trying to trigger the tests but it will not help if you got many pings within one hour. To save the resource, we need to do more investigation before blindly triggering the test.

**Comment by HyukjinKwon** [comment]
I am not blindly triggering the test. I skimmed and only re-triggered tests some PRs while I am skimming stale PRs.

For "ok to test", I haven't also blindly re-triggered. I did where any committer initially triggered or I see some values while skimming stale PRs.

Getting pings is annoying of course but the root cause is we didn't check the PRs so far diligently, or authors were being inactive.

**Comment by gatorsmile** [comment]
If you want to help this, we should ping the reviewers/committers based on the priority of these PRs. Also, we should not trigger the pings within a short period. Please reduce the size of the batch to 10s per day? 

We need to seriously consider which new features we really want to introduce in the upcoming release. We only have 15 days left before the code freeze of Spark 2.4.

**Comment by HyukjinKwon** [comment]
Of course, the time and priority matter but there are pending PRs queued up due to the time and priority matter so far.

Shall we check the PRs and see if there are important ones for Spark 2.4 before the branch is cut out since the release is being close? Otherwise, most of them will probably be missed by the time and priority matter again.

I don't think I have done ping and retriggering things for roughly about the last year in such batch.

**Comment by HyukjinKwon** [comment]
Also, the root cause of getting a lot of pings is that we somehow started to block the Jenkins build for some old PRs. Probably I missed some threads in dev mailing list but I still don't know who and who started this.

In a way to be honest, I kind of felt getting annoyed by this to somehow probably by the similar reason you gave, even though I haven't expressed this so far for my reasons above.

**Comment by gatorsmile** [comment]
@HyukjinKwon Currently, in my opinion, the highest priority PRs include Parquet nested column pruning (https://github.com/apache/spark/pull/21320/files), new built-in avro, high-order functions, data correctness issue of Shuffle+Repartition on RDD, gang scheduling and so on. If you have bandwidth, please also help the reviews in these PRs.

**Comment by cloud-fan** [comment]
@dbtsai I think this PR can be unblocked if we only turn `In` to `EqualTo` if the `list` is not a `ListQuery`. We can work around the type mismatch problem for now.

**Comment by dbtsai** [comment]
@HyukjinKwon thanks for bringing this to my attention. @gatorsmile I thought the bug is found by this PR, and not in this PR. This PR is blocked until SPARK-24443 is addressed. I'll unblocck this PR by turning `In` to `EqualTo` if the `list` is not a `ListQuery` suggested by @cloud-fan . Thanks all.

**Comment by viirya** [comment]
retest this please.

**Comment by gatorsmile** [comment]
LGTM. The test failure is not related to this PR. 

Thanks! Merged to master.

**Comment by viirya** [comment]
Seems like a related test failure? cc @dbtsai

**Comment by HyukjinKwon** [comment]
This hasn't get any test pass even once. The test is broken by this commit:

Before:

```
$ build/sbt "~sql/test-only *SQLQueryTestSuite -- -z not-in-unit-tests-multi-column-literal.sql"
...
[info] SQLQueryTestSuite:
[info] - subquery/in-subquery/not-in-unit-tests-multi-column-literal.sql (4 seconds, 602 milliseconds)
...
[info] All tests passed.
```

After:

```
$ build/sbt "~sql/test-only *SQLQueryTestSuite -- -z not-in-unit-tests-multi-column-literal.sql"
...
[info] SQLQueryTestSuite:
[info] - subquery/in-subquery/not-in-unit-tests-multi-column-literal.sql *** FAILED *** (4 seconds, 388 milliseconds)
[info]   Expected "struct<[a:int,b:decimal(2,1)]>", but got "struct<[]>" Schema did not match for query #1
[info]   -- Case 5
[info]     -- (one null column with no match -> row is returned)
[info]   SELECT *
[info]   FROM   m
[info]   WHERE  b = 1.0 -- Matches (null, 1.0)
[info]          AND (a, b) NOT IN ((2, 3.0)) (SQLQueryTestSuite.scala:246)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:528)
[info]   at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1560)
[info]   at org.scalatest.Assertions$class.assertResult(Assertions.scala:1003)
...
[error] Failed: Total 1, Failed 1, Errors 0, Passed 0
[error] Failed tests:
[error] 	org.apache.spark.sql.SQLQueryTestSuite
```

Reverting this since this will probably break the build and this PR hasn't got any successful build in any event.

**Comment by cloud-fan** [comment]
@dbtsai can you open a new PR? thanks!

**Comment by dbtsai** [comment]
I opened a new PR at https://github.com/apache/spark/pull/21797/files Will work on the test issue there. Thanks.

**Pull request closed without merging**
